src/core/abr/buffer_based_chooser.ts

```diff
@@ -0,0 +1,159 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Observable,
+} from "rxjs";
+import {
+  filter,
+  map,
+  tap,
+  withLatestFrom,
+} from "rxjs/operators";
+import {
+  ISegment,
+  Representation,
+} from "../../manifest";
+import arrayFind from "../../utils/array_find";
+import arrayFindIndex from "../../utils/array_find_index";
+import EWMA from "./ewma";
+
+export interface IAppendedSegment {
+  representation: Representation;
+  segment: ISegment;
+  bufferGap? : number;
+}
+
+interface IEstimate {
+  representation: Representation;
+  EWMA: EWMA;
+}
+
+/**
+ * Uses BOLA rule to choose the representation that should keep the buffer stable.
+ *
+ * This chooser is a hybrid solution that also relies on representations
+ * "maintainability". Each time a chunk is downloaded, from the ratio between
+ * the chunk duration and chunk's request time, we can assume that the representation
+ * is "maintanable" or not. If so, we may switch to a better quality, or conversely to
+ * a worse quality.
+ *
+ * The switch decision is taken thanks to BOLA Rule. From bufferGap and quality levels,
+ * buffer steps are computed to know when we should switch quality.
+ */
+export default function BufferBasedChooser(
+  appendedSegment$: Observable<IAppendedSegment>,
+  scoreData$: Observable<IEstimate>,
+  representations: Representation[]
+): Observable<Representation> {
+  /**
+   * Get minimum buffer we should keep ahead to pick this representation index.
+   * @param {number} index
+   * @returns {number}
+   */
+  function minBufferLevelForRepresentation(index: number): number {
+    const qBitrate = bitrates[index];
+    const qUtility = utilities[index];
+    let min = 0;
+    for (let i = index - 1; i >= 0; --i) {
+        if (utilities[i] < utilities[index]) {
+            const iBitrate = bitrates[i];
+            const iUtility = utilities[i];
+            const level = Vp *
+              (gp + (qBitrate * iUtility - iBitrate * qUtility) /
+                (qBitrate - iBitrate));
+            min = Math.max(min, level);
+        }
+    }
+    return min;
+  }
+
+  /**
+   * From the buffer gap, choose a representation.
+   * @param {number} bufferGap
+   * @returns {undefined|Object}
+   */
+  function getEstimate(
+    bufferGap: number,
+    scoreData: IEstimate,
+    lastChosenRepresentation?: Representation
+  ): Representation {
+      const currentScore = scoreData.EWMA.getEstimate();
+      const index = arrayFindIndex(representations, (r) => {
+        return r.id === (scoreData as IEstimate).representation.id;
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4325))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=442&issues=AWg4IEc2Si5GWQS9ds0y&open=AWg4IEc2Si5GWQS9ds0y)

[ISSUE_KEY]: AWg4IEc2Si5GWQS9ds0y